File: src/models/Scooter.js
/** this class is used to create a scooter object
 * 
 * @param {number} id - the id of the scooter
 * @param {string} tag - the tag of the scooter max 8 characters
 * @param {string} status - the status of the scooter can Be IDLE, IN_USE and MAINTENANCE
 * @param {object} gpsLocation - the gps location of the scooter storing longitude and latitude
 * @param {number} mileage - the mileage of the scooter in km
 * @param {number} batterCharge - the battery charge of the scooter in %
 * 
 * 
 * @returns {object} - a scooter object
 * @author Marco de Boer
 */


export class Scooter {
    id;
    tag;
    status;
    gpsLocation;
    mileage;
    batteryCharge;


   static Status = {
        IDLE: "IDLE",
        IN_USE: "IN_USE",
        MAINTENANCE: "MAINTENANCE"
    }
    
    constructor(id, tag, status, gpsLocation, mileage, batteryCharge) {
        this.id = id;
        this.tag = tag;
        this.status = status;
        this.gpsLocation = gpsLocation;
        this.mileage = mileage;
        this.batteryCharge = batteryCharge;
    }

    static async createSampleScooter(pId = 0){
        let id = pId;
        let tag = createRandomTag(8);
        let status = createRandomStatus();
        let gpsLocation = createRandomGPSLocation();
        let mileage = Math.floor(Math.random() * 10000);
        let batteryCharge = Math.floor(Math.random() * 95) + 5;
        
        return new Scooter(id, tag, status, gpsLocation, mileage, batteryCharge);
    }

    static cloneScooter(scooter){
       if (scooter === null || scooter === undefined){
           return null;
       }
       return Object.assign(new Scooter(0), scooter);
    }
    
}

/** This creates a random status and is used to create a sample scooter
 * 
 * @author Marco de Boer
 * @returns {string} - a random status from the Scooter.Status object
 */

function createRandomStatus(){
    let statusArray = [Scooter.Status.IDLE, Scooter.Status.IN_USE, Scooter.Status.MAINTENANCE];
    let randomIndex = Math.floor(Math.random() * 3);
    return statusArray[randomIndex];
}

/**
 * This creates a random gps location near Amsterdam and is used to create a sample scooter
 * 
 * @author Marco de Boer
 * @returns {object} - a gps location object with a random latitude and longitude near Amsterdam
 */
function createRandomGPSLocation(){
    let latitude = Math.round((Math.random() * (0.08) + 52.3000)* 10000) / 10000;
    let longitude = Math.round((Math.random() * (0.2) + 4.8) * 10000) / 10000;
    return {latitude: latitude, longitude: longitude};
}

/** This creates a random string with the given length and is used to create a sample scooter
 * 
 * @author From stackoverflow user csharptest.net
 * @param {number} length - the length of the random string
 * @returns {string} - a random string with the given length
 */
function createRandomTag(length) {
    let result = '';
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    const charactersLength = characters.length;
    let counter = 0;
    while (counter < length) {
      result += characters.charAt(Math.floor(Math.random() * charactersLength));
      counter += 1;
    }
    return result;
}

File: es-server/src/main/java/app/models/Scooter.java
package app.models;

import com.fasterxml.jackson.annotation.JsonView;
import jakarta.persistence.*;
import org.hibernate.annotations.OnDelete;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Set;

@Entity
@Table(name= "Scooter")
@NamedQuery(name = "Scooter_find_by_status", query = "SELECT s FROM Scooter s WHERE s.status = ?1")
@NamedQuery(name = "Scooter_find_by_battery", query = "SELECT s FROM Scooter s WHERE s.batteryCharge <= ?1")
public class Scooter implements Identifiable {
    @JsonView(Views.Summary.class)
    @Column(name = "tag")
    private String tag;
    @JsonView(Views.Summary.class)
    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private Status status;

    public void setGpsLocation(GPSLocation gpsLocation) {
        this.gpsLocation = gpsLocation;
    }

    @JsonView(Views.Summary.class)
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id")
    private long id;

    @OneToOne(mappedBy = "scooter", cascade = CascadeType.ALL)
    private GPSLocation gpsLocation;


    @Column(name = "mileage")
    private int mileage;
    @JsonView(Views.Summary.class) private int batteryCharge;

    @OneToMany(mappedBy = "scooter", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Trip> trips = new ArrayList<>();

    public Scooter() {

    }

    public boolean associateTrip(Trip trip){
        //Check if already associated and do nothing.
        if (trips.contains(trip)){
            return false;
            // If trip is not associated add to the list and change it in the trip aswell.
        } else {
            trips.add(trip);
            trip.associateScooter(this);
            return true;
        }
    }

    public boolean dissociateTrip(Trip trip){
        if (trips.contains(trip)){
            trips.remove(trip);
            return true;
        } else {
            return false;
        }
    }

    public class Views {
        public static class Summary {

        }

    }


    public List<Trip> getTrips() {
        return trips;
    }

    public void setTrips(List<Trip> trips) {
        this.trips = trips;
    }

    /**
     *
     * This constructor is used when parsing json to a new scooter instance.
     */

    public Scooter(long id, String tag, Status status, GPSLocation gpsLocation, int mileage, int batteryCharge) {
        this.id = id;
        this.tag = tag;
        this.status = status;
        this.gpsLocation = gpsLocation;
        this.mileage = mileage;
        this.batteryCharge = batteryCharge;
    }

    public enum Status {
        IDLE, IN_USE, MAINTENANCE, UNAVAILABLE;

        private static final Random PRNG = new Random();

        public static Status randomStatus(){
            Status[] statusses = values();
            return statusses[PRNG.nextInt(statusses.length)];
        }
    }

    public Scooter(String tag) {
        this.tag = tag;
    }

    public Scooter(long id) {
        this.id = id;
    }

    public static Scooter createSampleScooter(long id){
        Scooter scooter = new Scooter(id);
        scooter.setTag(createRandomTag(8));
        scooter.setStatus(Status.randomStatus());
        scooter.setGPSLocation(GPSLocation.createRandomGPSLocation());
        scooter.setMileage((int) Math.round(Math.random() * 10000));
        scooter.setBatteryCharge((int) Math.floor(Math.random() * 95) + 5);
        return scooter;
    }

    private static String createRandomTag(int length) {
        String result = "";
        final String characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        final int charactersLength = characters.length();
        int counter = 0;
        while (counter < length) {
            result += characters.charAt((int) Math.floor(Math.random() * charactersLength));
            counter += 1;
        }
        return result;
    }

    @Override
    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public void setTag(String tag) {
        this.tag = tag;
    }

    public void setStatus(Status status) {
        this.status = status;
    }

    public void setGPSLocation(GPSLocation gpsLocation) {
        this.gpsLocation = gpsLocation;
    }

    public void setMileage(int mileage) {
        this.mileage = mileage;
    }

    public void setBatteryCharge(int batteryCharge) {
        this.batteryCharge = batteryCharge;
    }

    public String getTag() {
        return tag;
    }

    public Status getStatus() {
        return status;
    }

    public GPSLocation getGpsLocation() {
        return gpsLocation;
    }

    public int getMileage() {
        return mileage;
    }

    public int getBatteryCharge() {
        return batteryCharge;
    }


    @Override
    public String toString() {
        return "Scooter{" +
                "id=" + id +
                ", tag='" + tag + '\'' +
                ", status='" + status + '\'' +
                ", gpsLocation='" + gpsLocation + '\'' +
                ", mileage=" + mileage +
                ", batteryCharge=" + batteryCharge +
                '}';
    }


}
